common/src/main/java/org/apache/drill/common/Typifier.java

```diff
@@ -88,6 +96,40 @@ public class Typifier {
   // If a String contains any of these, try to evaluate it as an equation
   private static final char[] MathCharacters = new char[]{'+', '-', '/', '*', '='};
 
+  /**
+   * This function infers the Drill data type of unknown data.
+   * @param data The input text of unknown data type.
+   * @return A {@link MinorType} of the Drill data type.
+   */
+  public static MinorType typifyToDrill (String data) {
+    Entry<Class, String> result = Typifier.typify(data);
+    String dataType = result.getKey().getSimpleName();
+
+    // If the string is empty, return UNKNOWN
```

Comment:
The next line of code contradicts this comment by returning VARCHAR. 
(Unless VARCHAR == UNKNOWN, which is news to me.)

Reply:
@mbeckerle Drill doesn't really have an `UNKNOWN` data type.   The way the typifier works is that if it can't determine the datatype, it falls back to string which can basically accept anything.

Regarding the lists...  The issue is that to create a list, you have to set the data mode to `REPEATED`.  The problem with XML is that there's no real way to know if a field is repeated or not.  Consider this:

```xml

<book>
  <author>a</author>
</book>
<book>
    <author>a1</author>
    <author>a2</author>
</book>
```

Since Drill uses the streaming reader, when it first encounters the `author` field, it would add an entry for a VARCHAR field.  However, when it gets to the next author record, it should be list, but there's no way to really know that w/o a schema.  

With JSON we don't have this problem because it uses `[` to denote lists. 
 
Does that make sense?

Reply:
Makes perfect sense. 

For XML you need XSD to know what's potentially repeating. 

Sometimes that is easy because of minOccurs/maxOccurs.

But there's also these "implied arrays".
```
<element name="a" type="xs:int"/><!-- this is a[1] -->
<element name="b" type="xs:int"/>
<element name="a" type="xs:int"/><!-- this is a[2] -->
```
That's allowed in both XSD and DFDL schemas (though I want to change Daffodil to issue a warning if you do this, because it is such a bad idea when representing structured data.)

The element 'a' looks like an array, in that you can index it. 

I think for drill there are just 2 columns: 'a', 'b',  but as there is more than one declaration for 'a', it is an implied array. 

Even just detecting this (and disallowing it for now) requires a more sophisticated metadata builder which is what I'm working on now.